components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +337,159 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
+        if month > max_month {
+            return (
+                Err(CalendarError::Overflow {
+                    field: "month",
+                    max: max_month as usize,
+                }),
+                cache,
+            );
+        }
+
+        let max_day_tuple = Self::days_in_month_cached(year, month, Some(cache));
+        let max_day = max_day_tuple.0;
+        if day > max_day {
+            return (
+                Err(CalendarError::Overflow {
+                    field: "day",
+                    max: max_day as usize,
+                }),
+                cache,
+            );
+        }
+
+        // Unchecked can be used because month and day are already checked in this fn
+        (
+            Ok(ArithmeticDate::<C>::new_unchecked(year, month, day)),
+            cache,
+        )
+    }
+
+    /// Call `months_in_year_cached` on a `ChineseBasedDateInner`
+    pub(crate) fn months_in_year_inner(&self) -> u8 {
+        Self::months_in_year_cached(self.0.year, Some(self.1)).0
+    }
+
+    /// Return the number of months in a given year, which is 13 in a leap year, and 12 in a common year.
+    /// Also returns a ChineseBasedCache - either the cache passed as an argument, if applicable, or a new
+    /// ChineseBasedCache for the given year.
+    fn months_in_year_cached(
+        year: i32,
+        cache_arg: Option<ChineseBasedCache>,
+    ) -> (u8, ChineseBasedCache) {
+        let (is_leap_year, cache) = Self::is_leap_year_cached(year, cache_arg);
+        (if is_leap_year { 13 } else { 12 }, cache)
+    }
+
+    /// Returns true if the given year is a leap year, and false if not; also returns a ChineseBasedCache -
+    /// either the cache passed as an argument, if applicable, or a new ChineseBasedCache for the given year.
+    fn is_leap_year_cached(
+        year: i32,
+        cache_arg: Option<ChineseBasedCache>,
+    ) -> (bool, ChineseBasedCache) {
+        if let Some(existing_cache) = cache_arg {
+            (existing_cache.is_leap_year, existing_cache)
+        } else {
+            let mid_year = Self::fixed_mid_year_from_year(year);
+            Self::fixed_date_is_in_leap_year_empty_cache(mid_year)
+        }
+    }
+
+    /// Returns true if the given fixed date is in a leap year; also returns a reference
+    /// to a new mutable ChineseBasedCache.
+    fn fixed_date_is_in_leap_year_empty_cache(date: RataDie) -> (bool, ChineseBasedCache) {
+        let (new_year, solstice) = Self::new_year_on_or_before_fixed_date(date, None);
+        let is_leap_year = Self::new_year_is_leap_year(new_year, Some(solstice));
+        let cache = ChineseBasedCache {
+            new_year,
+            is_leap_year,
+        };
+        (is_leap_year, cache)
+    }
+
+    /// Gets the leap month for an instance of `ChineseBasedDateInner`; this function assumes the `ChineseBasedDateInner`
+    /// represents a year with a leap month.
+    pub(crate) fn get_leap_month(&self) -> u8 {
```

Comment:
thought: i feel like it _may_ be worth caching which month is the leap month so that month code calculation works? we currently don't store the month code? Getting the month code is going to be a common operation and we want formatting to be a happy path.

Reply:
not really sure here, but I don't like us calling `no_major_solar_term` multiple times in the month code getter

Reply:
either way can be thought about later

Reply:
This would be something good for hard-coding, but calculating which month in a year is the leap month is an expensive operation that is not normally done in date creation; I chose to put `new_year` and `is_leap_year` in the cache because they are useful and already calculated/needed a few times in all the ways a date can be created (from iso, from codes, or `Date::try_new`). But it might be worth the higher initial cost if a caller uses `.month()` multiple times, so it doesn't need to be recalculated.

Reply:
sounds good.

Reply:
`ChineseBasedCache` now stores a `leap_month: u8` field which is either 1..=13 if `is_leap_year` is true or 14 otherwise, as was done before (except now that it is stored instead of computed). I ran the benchmark test and it makes it around 1 or 2 microseconds slower, but that is not very much change proportionally (~47 microseconds -> ~49), and I think this will be very beneficial for situations where `month()` is called frequently on the same `Date<Chinese>` or `Date<Dangi>`. Also, this will make hard-coded values very easy since we planned to have this anyway.